packages/api/src/agent.ts

```diff
@@ -159,23 +159,45 @@ export class AtpAgent {
     try {
       this.session = session
       const res = await this.api.com.atproto.server.getSession()
-      if (!res.success || res.data.did !== this.session.did) {
-        throw new Error('Invalid session')
+      if (res.data.did !== this.session.did) {
+        throw new XRPCError(
+          ResponseType.InvalidRequest,
+          'Invalid session',
+          'InvalidDID',
+        )
       }
       this.session.email = res.data.email
       this.session.handle = res.data.handle
       this.session.emailConfirmed = res.data.emailConfirmed
       this._updateApiEndpoint(res.data.didDoc)
+      this._persistSession?.('update', this.session)
       return res
     } catch (e) {
       this.session = undefined
-      throw e
-    } finally {
-      if (this.session) {
-        this._persistSession?.('create', this.session)
+
+      if (e instanceof XRPCError && e.error) {
+        /*
+         * `ExpiredToken` and `InvalidToken` are handled in
+         * `this_refreshSession`, and emit an `expired` event there.
+         *
+         * `AuthMissing` is a respose from `getSession`, and happens AFTER the
+         * initial `expired` event is sent. If we handle that here, we'll
+         * double-emit `expired`.
+         *
+         * `InvalidDID` is emit above, and should be handled here.
+         *
+         * Everything else is unexpected.
+         */
+        if (['InvalidDID'].includes(e.error)) {
+          this._persistSession?.('expired', undefined)
+        } else if (['InternalServerError'].includes(e.error)) {
+          this._persistSession?.('network-error', undefined)
+        }
```

Comment:
I was under the impression that network errors (e.g. user does not have connectivity to the service) would have status `ResponseType.Unknown`, and may not have an `error` property.

Reply:
Alright, I think I see!  These 500s are treated akin to a network error.  That makes sense since it's another kind of temporary unavailability, and we don't necessarily want to log the user out.  Other status codes for temporary unavailability such as 502 and 503 might be handled similarly.

Tiny P.S.: I think that checking the `status` property of an `XRPCError` for this kind of thing might be more reliable.

Reply:
`network-error` is just a name Paul suggested for this. This particular branch is tested in `agent.test.ts`, where we do get an `XRPCError` called `InternalServerError`.

Basically we're trying to handle any not-already-handled knowns with `expired`, and leave unknowns (like a network timeout that doesn't even hit our server) as `network-error`.